src/sundials_matrix_wrapper.cpp

```diff
@@ -219,6 +219,50 @@ void SUNMatrixWrapper::multiply(gsl::span<realtype> c, gsl::span<const realtype>
     }
 }
 
+void SUNMatrixWrapper::sparse_multiply(AmiVectorArray C,
+                                       SUNMatrixWrapper B,
+                                       gsl::span<int> colsB) const {
+    if (!matrix)
+        return;
+    
+    if (colsB.size() < 1)
+        return;
+    
+    sunindextype nrows = rows();
+    sunindextype ncols = columns();
+
+    if (static_cast<sunindextype>(NV_LENGTH_S(C.getNVector(0))) != nrows)
+        throw std::invalid_argument("Dimension mismatch between number of rows "
+                                    "in A (" + std::to_string(nrows) + ") and "
+                                    "number of rows in C (" + std::to_string(NV_LENGTH_S(C.getNVector(0)))
+                                    + ")");
+
+    if (B.rows() != ncols)
+        throw std::invalid_argument("Dimension mismatch between number of cols "
+                                    "in A (" + std::to_string(ncols)
+                                    + ") and number of rows in B ("
+                                    + std::to_string((int)B.rows()) + ")");
+    
+    if (SUNMatGetID(matrix) != SUNMATRIX_SPARSE)
+        throw std::invalid_argument("Matrix A not sparse in sparse_multiply");
+    
+    if (sparsetype() != CSC_MAT)
+        throw std::invalid_argument("Matrix A not of type CSC_MAT");
+
+    if (SUNMatGetID(B.matrix) != SUNMATRIX_SPARSE)
+        throw std::invalid_argument("Matrix B not sparse in sparse_multiply");
+    
+    if (B.sparsetype() != CSC_MAT)
+        throw std::invalid_argument("Matrix B not of type CSC_MAT");
+
+    /* Carry out actual multiplication */
+    for (int icol = 0; icol < (int)colsB.size(); ++icol)
+        for(sunindextype k = B.indexptrs_ptr[colsB[icol]]; k < B.indexptrs_ptr[colsB[icol] + 1]; ++k)
+            for(sunindextype l = indexptrs_ptr[k]; l < B.indexptrs_ptr[k + 1]; ++l)
+                C.data(icol)[l] += data_ptr[l] * B.data_ptr[k];
```

Comment:
Saving `C.data(icol)` in outmost loop and access via pointer in here may save a little time.

Reply:
Will be changed anayhow, as C will no longer be a VectorArray, but a sparse matrix soon...